File: 1261.Find_Elements_in_a_Contaminated_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 *struct TreeNode {
 *    int val;
 *    TreeNode *left;
 *    TreeNode *right;
 *    TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
  };
 *
 * Goal: Return true if the target value exists in the recovered binary tree.
 * Intuition: DFS to recover the tree and insert into a set, so that we can know
 * it in O(1)
 * Time Complexity: O(n)
 */

#include <unordered_set>
class FindElements
{
  private:
    std::unordered_set<int> values;
    void dfs(const int &parentVal, TreeNode *node)
    {
        if (node == nullptr)
            return;

        values.insert(parentVal);

        dfs(parentVal * 2 + 1, node->left);
        dfs(parentVal * 2 + 2, node->right);
    }

  public:
    FindElements(TreeNode *root)
    {
        dfs(0, root);
    }

    bool find(int target)
    {
        return values.count(target);
    }
};

/**
 * Your FindElements object will be instantiated and called as such:
 * FindElements* obj = new FindElements(root);
 * bool param_1 = obj->find(target);
 */

File: 1261.Find_Elements_in_a_Contaminated_Binary_Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type FindElements struct {
	vals map[int]bool
}

func Constructor(root *TreeNode) FindElements {
	FE := FindElements{vals: make(map[int]bool)}
	dfs(root, 0, &FE)
	return FE
}

func dfs(node *TreeNode, currentVal int, FE *FindElements) {
	if node == nil {
		return
	}

	FE.vals[currentVal] = true
	dfs(node.Left, currentVal*2+1, FE)
	dfs(node.Right, currentVal*2+2, FE)
}

func (this *FindElements) Find(target int) bool {
	return this.vals[target]
}

/**
 * Your FindElements object will be instantiated and called as such:
 * obj := Constructor(root);
 * param_1 := obj.Find(target);
 */
